### PR TITLE
16972 localize api v3

### PR DESCRIPTION
**Pull request by hschink**
https://community.openproject.org/work_packages/16972
https://community.openproject.org/work_packages/17320
https://community.openproject.org/work_packages/17582

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://fix-16972-localize-api-v3.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://fix-16972-localize-api-v3.opf-openproject-5eefc6be83fa70526458.ttrcloud.com' target='_blank'>Click to open.</a>

File: lib/api/errors/unwritable_property.rb [hunk 1]
```diff
@@ -34,8 +34,10 @@ def initialize(invalid_attributes)
         attributes = Array(invalid_attributes)
 
         fail ArgumentError, 'UnwritableProperty error must contain at least one invalid attribute!' if attributes.empty?
+        message = I18n.t('api_v3.errors.multiple_errors')
+        message = I18n.t('api_v3.errors.writing_read_only_attributes') if attributes.length == 1
```
Comment (ulferts) [inline lib/api/errors/unwritable_property.rb]: I'd prefer a `if else` here

File: config/locales/en.yml [hunk 2]
```diff
@@ -233,7 +233,7 @@ en:
               not_a_valid_parent: "is invalid"
               does_not_exist: "does not exist"
             status_id:
-              status_transition_invalid: "no valid transition exists from old to new status for the current user roles"
+              status_transition_invalid: "is invalid because no valid transition exists from old to new status for the current user roles"
```
Comment (ulferts) [inline config/locales/en.yml]: Couldn't this be better phrased as "...for the current user's roles"

File: config/locales/de.yml [hunk 3]
```diff
@@ -231,7 +231,7 @@ de:
               not_a_valid_parent: "ist ungültig"
               does_not_exist: "existiert nicht"
             status_id:
-              status_transition_invalid: "es existiert kein valider Übergang vom alten zum neuen Status für die aktuellen Nutzer-Rollen"
+              status_transition_invalid: "ist ungültig da kein valider Übergang vom alten zum neuen Status für die aktuellen Nutzer-Rollen existiert"
```
Comment (ulferts) [inline config/locales/de.yml]: Please see english

**Comment by ulferts** [comment]
I like it that the `Accept-Language` header is used.

File: lib/api/errors/validation.rb [hunk 4]
```diff
@@ -30,10 +30,47 @@
...
   module Errors
     class Validation < ErrorBase
-      def initialize(obj)
-        messages = obj.respond_to?(:errors) ? obj.errors.full_messages : Array(obj)
+      def self.create(errors)
+        properties = errors.keys
 
-        super 422, (messages.length == 1) ? messages[0] + '.' : 'Multiple fields violated their constraints.'
+        # Merge property error messages (e.g. for status and status_id)
+        properties.each do |p|
```
Comment (ulferts) [inline lib/api/errors/validation.rb]: I always consider the requirement to comment code as a strong indication that the commented code should be placed in it's own method. Same here.

File: lib/api/errors/validation.rb [hunk 5]
```diff
@@ -30,10 +30,47 @@
...
+        end
+
+        errors.keys.each_with_object({}) do |key, hash|
+          messages = errors[key].each_with_object([]) do |m, l|
+            # Let's assume that standard validation errors never end with a
+            # punctuation mark. Then it should be fair enough to assume that we
+            # don't need to prepend the error key if the error ends with a
+            # punctuation mark. Let's hope that this is true for the languages
+            # we'll support in OpenProject.
+            if m =~ /(\.|\?|\!)\z/
```
Comment (ulferts) [inline lib/api/errors/validation.rb:43]: I haven't found a better solution yet but I don't like this a bit.

Comment (hschink) [inline lib/api/errors/validation.rb:43]: But even if a language doesn't support these western punctuations (see [Wikipedia](http://en.wikipedia.org/wiki/Punctuation#Other_languages) for different punctuation styles) we still return an error message.

Comment (ulferts) [inline lib/api/errors/validation.rb:43]: As discussed, this should be done differently, but doing it differently is out of scope for now. Merging this now and will hate myself later.

**Pull request merged**